aspnet/fundamentals/configuration.rst

```diff
@@ -2,169 +2,252 @@
+.. _options-config-objects:
 
-.. note:: To override nested keys through environment variables in shells that don't support ``:`` in variable names, replace them with ``__`` (double underscore).
+Using Options and configuration objects
+---------------------------------------
 
-It can be useful to have environment-specific configuration files. This can be achieved using the following:
+The options pattern enables using custom options classes to represent a group of related settings. We recommended that you create decoupled classes for each feature within your app. Decoupled classes follow:
 
-.. literalinclude:: /../common/samples/WebApplication1/src/WebApplication1/Startup.cs
-  :dedent: 8
-  :language: none
-  :lines: 20-35
-  :emphasize-lines: 6
+- The `Interface Segregation Principle (ISP) <http://deviq.com/interface-segregation-principle/>`_ : Classes depend only on the configuration settings they use.
+- `Separation of Concerns <http://deviq.com/separation-of-concerns/>`_ : Settings for disparate parts of your app are managed separately, and not dependent or coupled with one another.
+
+The options class must be non-abstract with a public parameterless constructor. For example:
 
-The :dn:iface:`~Microsoft.AspNetCore.Hosting.IHostingEnvironment` service is used to get the current environment. In the ``Development`` environment, the highlighted line of code above would look for a file named ``appsettings.Development.json`` and use its values, overriding any other values, if it's present. Learn more about :doc:`environments`.
+.. literalinclude:: configuration/sample/src/UsingOptions/Models/MyOptions.cs
+  :language: c#
+  :lines: 3-7
+  :dedent: 4
 
-When specifying files as configuration sources, you can optionally specify whether changes to the file should result in the settings being reloaded. This is configured by passing in a ``true`` value for the ``reloadOnChange`` parameter when calling :dn:method:`~Microsoft.Extensions.Configuration.JsonConfigurationExtensions.AddJsonFile` or similar file-based extension methods.
+.. _options-example:
 
-.. warning:: You should never store passwords or other sensitive data in configuration provider code or in plain text configuration files. You also shouldn't use production secrets in your development or test environments. Instead, such secrets should be specified outside the project tree, so they cannot be accidentally committed into the configuration provider repository. Learn more about :doc:`environments` and managing :doc:`/security/app-secrets`.
+In the code below, ``ConfigurationBuilder`` is initialized in the ``Startup`` class. In the ``ConfigureServices`` method, an ``IConfigurationRoot`` interface is passed to the to :dn:method:`~Microsoft.Extensions.DependencyInjection.OptionsServiceCollectionExtensions.Configure\<TOptions>`, which adds ``ConfigurationBuilder`` to the dependency injection container:
```

Comment:
Per the style guide I'm removing the description and moving all code explanations to code comments.

>  Why do we use IConfigurationRoot here but we used IConfiguration in the first snippet in the article?

Nice catch. It should be IConfigurationRoot  everywhere, we want the root of the configuration.  IConfiguration is the lower level interface that is specialized by `IConfigurationRoot` and `IConfigurationSection`